### PR TITLE
Support for storage providers other than Azure Storage

### PR DESCRIPTION
**Pull request by ConnorMcMahon**
Proof of concept for #652 with DurableTask.Emulator provider. Will eventually add work for #641 as well. 

Only added tests for the emulator provider to a few tests for now, as I didn't want to effectively double the amount of time it takes for the CI to run.

### CODE REVIEW

File: src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs [hunk 1]
```diff
@@ -344,97 +339,13 @@ protected internal virtual DurableOrchestrationClient GetClient(OrchestrationCli
...
-                        // as the host when possible to ensure we any send operations can be picked up
-                        // immediately instead of waiting for the next queue polling interval.
-                        innerClient = this.orchestrationService;
-                    }
-                    else
-                    {
-                        innerClient = new AzureStorageOrchestrationService(settings);
-                    }
-
+                    var innerClient = this.orchestrationServiceFactory.GetOrchestrationClient(attribute);
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableTaskExtension.cs]: I prefer we use actual class names in variable decorations when the right-hand expression doesn't already contain it. It makes the code a bit more understandable.

File: src/WebJobs.Extensions.DurableTask/IOrchestrationServiceFactory.cs [hunk 2]
```diff
@@ -0,0 +1,15 @@
+﻿using System;
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/IOrchestrationServiceFactory.cs]: This new file needs a copyright header

File: src/WebJobs.Extensions.DurableTask/Options/AzureStorageOptions.cs [hunk 3]
```diff
@@ -0,0 +1,122 @@
+﻿using System;
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/AzureStorageOptions.cs]: This new file and other new files all need copywrite headers.

File: src/WebJobs.Extensions.DurableTask/DurableTaskJobHostConfigurationExtensions.cs [hunk 4]
```diff
@@ -27,7 +26,8 @@ public static IWebJobsBuilder AddDurableTask(this IWebJobsBuilder builder)
 
             builder.AddExtension<DurableTaskExtension>()
                 .BindOptions<DurableTaskOptions>()
-                .Services.AddSingleton<IConnectionStringResolver, WebJobsConnectionStringProvider>();
+                .Services.AddSingleton<IConnectionStringResolver, WebJobsConnectionStringProvider>()
+                         .AddSingleton<IOrchestrationServiceFactory, OrchestrationServiceFactory>();
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/DurableTaskJobHostConfigurationExtensions.cs]: With this model, do you think it would be possible for customers to provide their own `IOrchestrationServiceFactory` in the future via our DI support?

File: src/WebJobs.Extensions.DurableTask/IOrchestrationServiceFactory.cs [hunk 5]
```diff
@@ -0,0 +1,15 @@
...
+using System.Collections.Generic;
+using System.Text;
+using DurableTask.Core;
+using Microsoft.Azure.WebJobs.Extensions.DurableTask.Options;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask
+{
+    public interface IOrchestrationServiceFactory
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/IOrchestrationServiceFactory.cs]: All public types and members need to be documented with XML doc comments. I believe StyleCop and/or the compiler should be flagging this. If it's too much work to do it now, let's open an issue to track adding these (or a generic issue to track fixing all warnings).

File: src/WebJobs.Extensions.DurableTask/Options/AzureStorageOptions.cs [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+        /// </summary>
+        /// <value>Maximum interval for polling control and work-item queues.</value>
+        public TimeSpan MaxQueuePollingInterval { get; set; } = TimeSpan.FromSeconds(30);
+
+
+        internal override void ValidateHubName(string hubName)
+        {
+            try
+            {
+                NameValidator.ValidateBlobName(hubName);
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/AzureStorageOptions.cs]: In your other PR, we're making it so that the task hub name has no default. In that case, should we be checking for a null-or-empty `hubName` here?

File: src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs [hunk 7]
```diff
@@ -27,46 +23,10 @@ public class DurableTaskOptions
...
-        public int PartitionCount { get; set; } = 4;
-
-        /// <summary>
-        /// Gets or sets the visibility timeout of dequeued control queue messages.
-        /// </summary>
-        /// <value>
-        /// A <c>TimeSpan</c> configured by the host. The default is 5 minutes.
-        /// </value>
-        public TimeSpan ControlQueueVisibilityTimeout { get; set; } = TimeSpan.FromMinutes(5);
+        /// TODO: move this back into DurableTaskOptions.cs
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: Is this `TODO` valid?

File: src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs [hunk 8]
```diff
@@ -244,27 +169,13 @@ public class DurableTaskOptions
...
-        /// <value>Maximum interval for polling control and work-item queues.</value>
-        public TimeSpan MaxQueuePollingInterval { get; set; } = TimeSpan.FromSeconds(30);
-
         // Used for mocking the lifecycle notification helper.
         internal HttpMessageHandler NotificationHandler { get; set; }
 
         internal string GetDebugString()
         {
             var sb = new StringBuilder(4096);
             sb.AppendLine("Initializing extension with the following settings:");
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: What is the plan to get the other Options types to participate in populating this debug string?

File: src/WebJobs.Extensions.DurableTask/Options/EmulatorStorageOptions.cs [hunk 9]
```diff
@@ -0,0 +1,10 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Options
+{
+    public class EmulatorStorageOptions : CommonStorageProviderOptions
+    {
+    }
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/EmulatorStorageOptions.cs]: Add a coment inside the class definition explaining that there are no settings.

File: src/WebJobs.Extensions.DurableTask/Options/StorageProviderOptions.cs [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+        internal CommonStorageProviderOptions GetConfiguredProvider()
+        {
+            if (this.configuredProvider == null)
+            {
+                var storageProviderOptions = new CommonStorageProviderOptions[] { this.AzureStorage, this.Emulator };
+                var activeProviders = storageProviderOptions.Where(provider => provider != null);
+                if (activeProviders.Count() != 1)
+                {
+                    throw new InvalidOperationException("Must have one and only one storage provider configured.");
```
Comment (cgillum) [inline src/WebJobs.Extensions.DurableTask/Options/StorageProviderOptions.cs]: I would rephrase this slightly: `There must be exactly one storage provider configured.`

File: src/WebJobs.Extensions.DurableTask/Options/EmulatorStorageOptions.cs (hunk 9 above)
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/Options/EmulatorStorageOptions.cs]: Instead I added a remark in the xml that there is currently no configuration available for the Emulator storage provider. If you still think a comment inside the class definition is necessary, I will add one.

File: src/WebJobs.Extensions.DurableTask/Options/AzureStorageOptions.cs (hunk 6 above)
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/Options/AzureStorageOptions.cs]: The implementations of `CommonStorageProviderOptions.ValidateHubName()` are just going to be the storage provider specific limitations on task name. I restored the null check in the `DurableTaskOptions.Validate()` method, as that is a universal requirement.

File: src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs (hunk 8 above)
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/Options/DurableTaskOptions.cs]: I think I have addressed this now.

File: src/WebJobs.Extensions.DurableTask/DurableTaskJobHostConfigurationExtensions.cs (hunk 4 above)
Comment (ConnorMcMahon) [reply src/WebJobs.Extensions.DurableTask/DurableTaskJobHostConfigurationExtensions.cs]: I'm not quite sure what our customer facing DI support will look like, and especially how they will interact with extensions. I don't see why this approach wouldn't allow customers to provide their own `IOrchestrationServiceFactory` though.

File: test/FunctionsV2/TestDataGenerator.cs [hunk 11]
```diff
@@ -0,0 +1,32 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+using System.Collections.Generic;
+
+namespace Microsoft.Azure.WebJobs.Extensions.DurableTask.Tests
+{
+    public class TestDataGenerator
```
Comment (cgillum) [inline test/FunctionsV2/TestDataGenerator.cs]: This is pretty cool. I did not know about this.

**Comment by cgillum** [comment]
Looks like this PR needs to be updated to fix some merge conflicts, most likely caused by Sebastian's changes.
